docs/contributor/adr/0010-geoblocking-draft.md

```diff
@@ -0,0 +1,387 @@
+
+```yaml
+apiVersion: geoblocking.kyma-project.io/v1alpha1
+kind: GeoBlocking
+metadata:
+  name: geoblocking
+  namespace: kyma-system
+spec:
+  gbService:
+    secret: "gb-secret"
+    tokenUrl: https://oauth.example.com/oauth2/token
+    ipRange:
+      url: https://lists.example.com/lists/some-list
+      refreshInSeconds: 3600
+    events:
+      url: https://events.example.com/events
+      lobId: "kyma"
+      productId: "some-product"
+      systemId: "some-system"
+```
+
+##### IP range allow/block list populated by customer
+```yaml
+apiVersion: geoblocking.kyma-project.io/v1alpha1
+kind: GeoBlocking
+metadata:
+    name: geoblocking
+    namespace: kyma-system
+spec:
+  ipRange:
+    configMap: "my-own-ip-range-list"
+```
+
+### Technical details
+
+#### Policy download optimization
+
+In order to reduce unnecessary updates of the list of blocked IP ranges the ip-auth service should use ETag mechanism, which is supported by the SAP internal service.
+
+The local copy (stored in a Config Map) should be used to optimize the amount of connections to the SAP internal service:
+- if the local copy contains a newer version (lastUpdateTime) -> load it
+- if the update check has been recently performed (now - lastUpdateCheckTime < refreshInterval) -> don't check it again
+- if the update check hasn't been recently performed (now - lastUpdateCheckTime >= refreshInterval) -> check for an update
+- if there is a newer version -> download, apply it, update a local copy and set lastUpdateCheckTime and lastUpdateTime to the current time
+- if there is no newer version -> update lastUpdateCheckTime time (so other pods may skip the check)
+- if there is a problem with checking for an update -> log a warning
+- if there is a problem with updating the local copy -> log a warning and use a new version anyway
+
+Pods should slightly randomize an update check time to benefit from the above optimization (minimize a risk of performing the check by multiple pods at the same time).
+
+#### Quick IP check
+
+The list of blocked IP ranges may be big (thousands of entries), so analyzing whether an IP address matches any IP range in the list is too time consuming.
+
+The list of IP ranges changes rarely, so it is better to build a data structure that supports quick comparision of IP addresses. The good candidate is a [radix tree](https://en.wikipedia.org/wiki/Radix_tree).
+
+#### Events retention
+
+In order to not cause unnecessary delays the access events should be sent asynchronously:
+- after making access decision the ip-auth should generate an event and store it in a memory queue
+- separate thread should grab events from the queue and send them to the SAP internal service.
+
+This approach may cause issues if SAP internal service responds slowly or does not respond at all, because events would be consuming memory. System availability is a key factor here, so it is acceptable to drop events in order to ensure stability (prevent out of memory issues).
```

Comment:
I added an additional section about 'Graceful shutdown / probes' about this. I think the reasonable approach is to implement graceful shutdown, so the events are not lost during 'normal' operation. But we can't avoid losing some events is pod is crashed. It would be a price for 'simplicity'.